video/out/wayland_common.c

```diff
@@ -1297,3 +1313,31 @@ bool vo_wayland_config (struct vo *vo, uint32_t flags)
 
     return true;
 }
+
+void vo_wayland_request_frame(struct vo *vo, void *data, vo_wayland_frame_cb cb)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+    wl->frame.data = data;
+    wl->frame.function = cb;
+    MP_DBG(wl, "restart frame callback\n");
+    frame_callback(wl, NULL, 0);
+}
+
+bool vo_wayland_wait_frame(struct vo *vo)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+
+    if (!wl->frame.callback)
+        return false;
+
+    // Reduce timeout if there are more than 3 pending waits.
+    // Most likely the mpv isn't receiving frame callbacks then.
+    // So don't timeout too long to avoid AV desync.
+    // 10ms matches the time specified in eglSwapInterval(0) mesa patch for wayland.
+    const int timeout = (wl->frame.pending > 3 ? 10 : 100); // ms
```

Comment:
The timeout logic seems a bit questionable. So every frame will still take 10 seconds, even if there are at least 4 frames queued up? I'd say these should just be dropped.

With display-sync, which just writes as many frames as it can, every frame should actually block - essentially that's the whole point of it: locking on display refresh rate by blocking and waiting until the next vsync each time.